File: backend/app/api/v1/suppliers.py
from typing import List

from fastapi import APIRouter, Depends, HTTPException, Response, status
from sqlmodel import Session, select

from app.db.session import get_session
from app.models.supplier import Supplier
from app.schemas.supplier import SupplierCreate, SupplierOut, SupplierUpdate


router = APIRouter()


@router.post("/", response_model=SupplierOut, status_code=status.HTTP_201_CREATED)
def create_supplier(supplier_in: SupplierCreate, db: Session = Depends(get_session)):
    supplier_data = supplier_in.model_dump(mode="json")
    supplier = Supplier(**supplier_data)
    db.add(supplier)
    db.commit()
    db.refresh(supplier)
    return supplier


@router.get("/", response_model=List[SupplierOut])
def list_suppliers(skip: int = 0, limit: int = 100, db: Session = Depends(get_session)):
    suppliers = db.exec(select(Supplier).offset(skip).limit(limit)).all()
    return suppliers


@router.get("/{supplier_id}", response_model=SupplierOut)
def get_supplier(supplier_id: int, db: Session = Depends(get_session)):
    supplier = db.get(Supplier, supplier_id)
    if not supplier:
        raise HTTPException(status_code=404, detail="Supplier not found")
    return supplier


@router.put("/{supplier_id}", response_model=SupplierOut)
def update_supplier(
    supplier_id: int,
    supplier_update: SupplierUpdate,
    db: Session = Depends(get_session),
):
    supplier = db.get(Supplier, supplier_id)
    if not supplier:
        raise HTTPException(status_code=404, detail="Supplier not found")

    update_data = supplier_update.model_dump(exclude_unset=True, mode="json")
    for field, value in update_data.items():
        setattr(supplier, field, value)

    db.add(supplier)
    db.commit()
    db.refresh(supplier)
    return supplier


@router.delete("/{supplier_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_supplier(supplier_id: int, db: Session = Depends(get_session)):
    supplier = db.get(Supplier, supplier_id)
    if not supplier:
        raise HTTPException(status_code=404, detail="Supplier not found")
    db.delete(supplier)
    db.commit()
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: frontend/lib/api/suppliers.ts
import { apiClient } from "./client";
import { Supplier, SupplierCreate, SupplierUpdate } from "@/types/supplier";
import { PaginationParams } from "@/types/api";

// Get all suppliers with pagination
export const getSuppliers = async (params?: PaginationParams): Promise<Supplier[]> => {
  const response = await apiClient.get<Supplier[]>("/api/v1/suppliers/", {
    params: {
      skip: params?.skip || 0,
      limit: params?.limit || 100,
    },
  });
  return response.data;
};

// Get supplier by ID
export const getSupplier = async (id: number): Promise<Supplier> => {
  const response = await apiClient.get<Supplier>(`/api/v1/suppliers/${id}`);
  return response.data;
};

// Create supplier
export const createSupplier = async (supplierData: SupplierCreate): Promise<Supplier> => {
  const response = await apiClient.post<Supplier>("/api/v1/suppliers/", supplierData);
  return response.data;
};

// Update supplier
export const updateSupplier = async (
  id: number,
  supplierData: SupplierUpdate
): Promise<Supplier> => {
  const response = await apiClient.put<Supplier>(`/api/v1/suppliers/${id}`, supplierData);
  return response.data;
};

// Delete supplier
export const deleteSupplier = async (id: number): Promise<void> => {
  await apiClient.delete(`/api/v1/suppliers/${id}`);
};

